compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc

```diff
@@ -6304,10 +6305,69 @@ INTERCEPTOR(int, fclose, __sanitizer_FILE *fp) {
 #endif
 
 #if SANITIZER_INTERCEPT_DLOPEN_DLCLOSE
+// Ordinary internal_readlink does not null-terminate strings.
+static int internal_readlink_with_null(const char *path, char *buf,
+                                       uptr bufsize) {
+  if (!buf) {
+    return -1;
+  }
+
+  // Ensure it is a valid string even if internal_readlink fails
+  buf[0] = '\0';
+
+  int len = internal_readlink(path, buf, bufsize);
+  if (len >= 0 && len < (int)bufsize)
+    buf[len] = '\0';
+
+  buf[bufsize - 1] = '\0';
+
+  return len;
+}
+
 INTERCEPTOR(void*, dlopen, const char *filename, int flag) {
   void *ctx;
   COMMON_INTERCEPTOR_ENTER_NOIGNORE(ctx, dlopen, filename, flag);
-  if (filename) COMMON_INTERCEPTOR_READ_STRING(ctx, filename, 0);
+
+  if (filename) {
+    COMMON_INTERCEPTOR_READ_STRING(ctx, filename, 0);
+
+#  if !SANITIZER_DYNAMIC
+    VPrintf(1, "dlopen interceptor: filename: %s\n", filename);
+
+    char *filename_canonical = (char *)InternalAlloc(sizeof(char) * 640);
+    if (filename_canonical)
+      internal_readlink_with_null(filename, filename_canonical, 640);
+    VPrintf(1, "dlopen interceptor: filename (canonical): %s\n",
+            filename_canonical);
+
+    // We care about dlopen(<main program>,...), which will only happen
+    // for statically linked ASan.
+    const char *self_fname;
+    int ret = dladdr_self_fname(&self_fname);
```

Comment:
Can you just compare dladdr_self_fname result with filename by pointer?

Reply:
Assuming we care about case from dladdr

Reply:
Ended up doing strcmp in case they do an intermediate strcpy